fe/fe-core/src/main/java/com/starrocks/sql/PlannerProfile.java

```diff
@@ -0,0 +1,133 @@
+// This file is made available under Elastic License 2.0.
+
+package com.starrocks.sql;
+
+import com.google.common.base.Preconditions;
+import com.starrocks.common.util.RuntimeProfile;
+import com.starrocks.qe.ConnectContext;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PlannerProfile {
+    private ConnectContext ctx;
+
+    public static class ScopedTimer implements AutoCloseable {
+        private long startTime = 0;
+        private volatile long currentThreadId = 0;
+        private long totalTime = 0;
+        private int totalCount = 0;
+        // possible to record p99?
+
+        public void start() {
+            Preconditions.checkState(currentThreadId == 0);
+            currentThreadId = Thread.currentThread().getId();
+            startTime = System.currentTimeMillis();
+        }
+
+        public void close() {
+            Preconditions.checkState(currentThreadId == Thread.currentThread().getId());
+            currentThreadId = 0;
+            totalTime += (System.currentTimeMillis() - startTime);
+            totalCount += 1;
+        }
+
+        public long getTotalTime() {
+            return totalTime;
+        }
+
+        public int getTotalCount() {
+            return totalCount;
+        }
+    }
+
+    private final HashMap<String, ScopedTimer> timers;
+
+    public PlannerProfile() {
+        timers = new HashMap<>();
+    }
+
+    public void init(ConnectContext ctx) {
+        this.ctx = ctx;
+    }
+
+    private synchronized ScopedTimer getOrCreateScopedTimer(String name) {
+        if (timers.containsKey(name)) {
+            return timers.get(name);
+        }
+        ScopedTimer t = new ScopedTimer();
+        timers.put(name, t);
+        return t;
+    }
+
+    private static final PlannerProfile DEFAULT_INSTANCE = new PlannerProfile();
+
+    public static ScopedTimer getScopedTimer(String name) {
+        // to avoid null.
+        PlannerProfile p = DEFAULT_INSTANCE;
+        ConnectContext ctx = ConnectContext.get();
+        if (ctx != null) {
+            p = ctx.getPlannerProfile();
+        }
+        ScopedTimer t = p.getOrCreateScopedTimer(name);
+        t.start();
+        return t;
+    }
+
+    private RuntimeProfile getRuntimeProfile(RuntimeProfile top, Map<String, RuntimeProfile> cache,
+                                             String prefix) {
+        if (cache.containsKey(prefix)) {
+            return cache.get(prefix);
+        }
+        String[] ss = prefix.split("\\.");
```

Comment:
Is it too tricky to use '.' to divide the hierarchy, can it be done in an explicit way?

Reply:
Any suggestion? Any character will do, but should be consistent in whole system.

Reply:
> Any suggestion? Any character will do, but should be consistent in whole system.

I haven't found any more elegant way, the explicit method will definitely cause more changes in other modules

Reply:
PlannerProfile.getScopedTimer(String

Reply:
Here is my suggestion

Replace `PlannerProfile.getScopedTimer(String name)` with `PlannerProfile.getScopedTimer(CascadeName name)`, class `CascadeName` encapsulate multi-level names.

In this way, name level is explicitly introduced rather than rely on `.` or some any other character

Reply:
Thanks. What I concern here is efficiency, string constant is more efficient.